File: Paths in Matrix Whose Sum Is Divisible by K.py
class Solution:
    def numberOfPaths(self, grid: List[List[int]], k: int) -> int:
        MOD = 10 ** 9 + 7
        n = len(grid)
        m = len(grid[0])
        dp = {}
        dp[(0, 0, grid[0][0] % k)] = 1

        for i in range(n):
            for j in range(m):
                for h in range(k):
                    if (i, j, h) not in dp:
                        dp[(i, j, h)] = 0

                    h2 = (h + grid[i][j]) % k

                    if (i, j, h2) not in dp:
                        dp[(i, j, h2)] = 0

                    if i - 1 >= 0:
                        dp[(i, j, h2)] += dp[(i - 1, j, h)]

                    if j - 1 >= 0:
                        dp[(i, j, h2)] += dp[(i, j - 1, h)]

                    dp[(i, j, h2)] %= MOD

        return dp[(n - 1, m - 1, 0)]

File: Paths in Matrix Whose Sum Is Divisible by K.cpp
class Solution {
public:
    int numberOfPaths(vector<vector<int>>& grid, int k) {
        constexpr int mod = 1e9 + 7;
        int m = grid.size();
        int n = grid[0].size();
        vector<vector<vector<int>>> dp(m, vector<vector<int>>(n, vector<int>(k, 0)));
        int r = grid[0][0] % k;
        dp[0][0][r] = 1;
        int prev = r;
        for (int i{1}; i < m; ++i) {
            int new_r = (prev+grid[i][0]) % k;
            dp[i][0][new_r] = 1;
            prev = new_r;
        }

        prev = r;
        for (int j{1}; j < n; ++j) {
            int new_r = (prev+grid[0][j]) % k;
            dp[0][j][new_r] = 1;
            prev = new_r;
        }

        for (int i{1}; i < m; ++i) {
            for (int j{1}; j < n; ++j) {
                int val = grid[i][j] % k;
                for (int k2{0}; k2 < k; ++k2) {
                    int offset = (k-val+k2) % k;
                    dp[i][j][k2] += dp[i-1][j][offset] + dp[i][j-1][offset];
                    dp[i][j][k2] %= mod;
                }
            }
        }
        return dp[m-1][n-1][0];
    }
};
